File: prep/implementation/reference-implementation/MergeSort.java
import java.util.*;

public class MergeSort {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int n = in.nextInt();
      int[] a = new int[n];

      for (int i = 0; i < n; i++) {
        a[i] = in.nextInt();
      }

      display(a, n);
      sort(a, n);
      display(a, n);
    }
  }

  private static void display(int[] a, int n) {
    for (int i = 0; i < n; i++) {
      System.out.printf("%d ", a[i]);
    }
    System.out.println();
  }

  private static void sort(int[] a, int n) { sort(a, 0, n - 1); }

  private static void sort(int[] a, int low, int high) {
    if (low >= high) {
      return;
    }

    int mid = low + (high - low) / 2;
    sort(a, low, mid);
    sort(a, mid + 1, high);
    merge(a, low, mid, high);
  }

  private static void merge(int[] a, int low, int mid, int high) {
    int llen = mid - low + 1;
    int[] left = new int[llen];

    int rlen = high - mid;
    int[] right = new int[rlen];

    for (int i = low; i <= mid; i++) {
      left[i - low] = a[i];
    }

    for (int j = mid + 1; j <= high; j++) {
      right[j - mid - 1] = a[j];
    }

    int lpos = 0, rpos = 0;
    for (int k = low; k <= high; k++) {
      if (lpos >= llen) {
        a[k] = right[rpos++];
      } else if (rpos >= rlen) {
        a[k] = left[lpos++];
      } else if (left[lpos] <= right[rpos]) {
        a[k] = left[lpos++];
      } else {
        a[k] = right[rpos++];
      }
    }
  }
}

File: lc/coreprep/implementation/main_iterations/iteration_cpp/merge_sort.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

void display(const vector<int> &a) {
  for (int e : a) {
    cout << e << " ";
  }
  cout << "\n";
}

void merge(vector<int> &a, int low, int mid, int high) {
  if (low > mid || mid >= high) {
    return;
  }

  int llen = mid - low + 1;
  vector<int> left(llen);
  for (int i = 0; i < llen; i++) {
    left[i] = a[low + i];
  }

  int rlen = high - mid;
  vector<int> right(rlen);
  for (int i = 0; i < rlen; i++) {
    right[i] = a[mid + i + 1];
  }

  int lpos = 0, rpos = 0;
  for (int i = low; i <= high; i++) {
    if (lpos < llen && rpos < rlen) {
      if (left[lpos] <= right[rpos]) {
        a[i] = left[lpos++];
      } else {
        a[i] = right[rpos++];
      }
    } else if (lpos < llen) {
      a[i] = left[lpos++];
    } else {
      a[i] = right[rpos++];
    }
  }
}

void sort(vector<int> &a, int low, int high) {
  if (low >= high) {
    return;
  }

  int mid = low + (high - low) / 2;
  sort(a, low, mid);
  sort(a, mid + 1, high);
  merge(a, low, mid, high);
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int n;
  cin >> n;

  vector<int> a(n);
  for (int i = 0; i < n; i++) {
    cin >> a[i];
  }

  display(a);
  sort(a, 0, n - 1);
  display(a);

  return 0;
}